bigtable/internal/transport/connpool.go

```diff
@@ -108,123 +483,298 @@ func (p *BigtableChannelPool) Close() error {
+			select {
+			case <-probeCtx.Done():
+				e.health.addProbeResult(false, cfg.WindowDuration) // Count as failure if context is done
+				return
+			default:
+			}
+			// don't update conn  isAlts used entry for now.
+			_, err := e.conn.Prime(probeCtx)
+			e.health.addProbeResult(err == nil, cfg.WindowDuration)
+		}(entry, hcConfig)
+	}
+	wg.Wait()
+}
+
+// detectAndEvictUnhealthy checks connection health and evicts the worst unhealthy one if allowed.
+func (p *BigtableChannelPool) detectAndEvictUnhealthy(hcConfig btopt.HealthCheckConfig, allowEviction func() bool, recordEviction func()) bool {
+	if !allowEviction() {
+		return false // Too soon since the last eviction.
+	}
+
+	conns := p.getConns()
+	numConns := len(conns)
+	if numConns == 0 {
+		return false
+	}
+
+	var unhealthyIndices []int
+	for i, entry := range conns {
+		if !entry.health.isHealthy(hcConfig.MinProbesForEval, hcConfig.FailurePercentThresh) { // isHealthy() locks internally
+			unhealthyIndices = append(unhealthyIndices, i)
+		}
+	}
+
+	if len(unhealthyIndices) == 0 {
+		return false // All connections are healthy.
+	}
+
+	unhealthyPercent := float64(len(unhealthyIndices)) / float64(numConns) * 100.0
+	if unhealthyPercent >= float64(hcConfig.PoolwideBadThreshPercent) {
+		btopt.Debugf(p.logger, "bigtable_connpool: Circuit breaker tripped, %d%% unhealthy, not evicting\n", int(unhealthyPercent))
+		return false // Too many unhealthy connections, don't evict.
+	}
+
+	// Find the connection with the most failed probes among the unhealthy ones.
+	var worstEntry *connEntry
+	maxFailed := -1
+	for _, idx := range unhealthyIndices {
+		entry := conns[idx]                      // Safe, using snapshot
+		failed := entry.health.getFailedProbes() // getFailedProbes() locks internally
+		if failed > maxFailed {
+			maxFailed = failed
+			worstEntry = entry
+		}
+	}
+
+	if worstEntry != nil {
+		recordEviction() // Record eviction time *before* replacing. // Record eviction time *before* replacing.
+		p.replaceConnection(worstEntry)
+		return true // Eviction happened
+	}
+
+	return false
+}
+
+// replaceConnection closes the connection for the oldEntry
+func (p *BigtableChannelPool) replaceConnection(oldEntry *connEntry) {
+	p.dialMu.Lock() // Serialize replacements
+	defer p.dialMu.Unlock()
+
+	// Mark the connection
+	// if it is marked,
+	// it means another routine (health eviction or dynamic scale down) took over it.
+	if !oldEntry.markAsDraining() {
+		return
+	}
+
+	currentConns := p.getConns()
+	idx := -1
+	for i, entry := range currentConns {
+		if entry == oldEntry {
+			idx = i
+			break
+		}
+	}
+
+	// If the connection isn't in the slice, it was already removed.
+	// The drain process should still be kicked off.
+	if idx == -1 {
+		btopt.Debugf(p.logger, "bigtable_connpool: Connection to replace was already removed. Draining it.")
+		// thread safe to call waitForDrainAndClose as conn.Close() can be called multiple times.
+		go p.waitForDrainAndClose(oldEntry)
```

Comment:
will this ever happen? wouldn't oldEntry.markAsDraining() means it should already be drained by smoething else?